File: Solutions/3-Longest_Substring_Without_Repeating_Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == None:
            return 0

        if len(s) == 0:
            return 0

        if len(s) == 1:
            return 1
        
        result = []
        temp = s[0]
        for i in s[1:]:
            if i not in temp:
                temp += i
            elif i == temp[0]:
                temp = temp[1:] + i
            elif i == temp[-1]:
                result.append(temp)
                temp = i
            else:
                result.append(temp)
                temp = temp[temp.find(i) + 1:] + i
            result.append(temp)
        
        return len(max(result, key=len))

# Second fastest
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        result = 0
        sub_str = ""

        for item in s:
            if item not in sub_str:
                sub_str += item
            else:
                result = max(result, len(sub_str))
                sub_str = sub_str[sub_str.index(item) + 1:] + item

        result = max(result, len(sub_str))

        return result

# Fastest
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        j = -1
        chars = {}
        ans = 0
        for i,ch in enumerate(s):
            if ch in chars and chars[ch] > j:
                j = chars[ch]
            else:
                if ans < i-j:
                    ans = i-j
            chars[ch] = i
        return ans

File: Solutions/3-Longest_Substring_Without_Repeating_Characters/solution.rs
use std::cmp;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut max_len: i32 = 0;
        let mut head: i32 = 0;
        let mut index: Vec<i32> = vec![0; 128];
        
        for (j, c) in s.chars().enumerate() {
            let tail = j as i32;
            let i = (c as u32) as usize;
            head = cmp::max(index[i], head);
            max_len = cmp::max(max_len, tail + 1 - head);
            index[i] = (tail + 1) as i32;
        }
        max_len
    }
}

// Another fastest
use std::cmp;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut sequence_start = 0;
        let mut max_sequence_len = 0;
        
        for current_index in 0..s.len() { 
            let sequence = &s[sequence_start..current_index];
            
            match sequence.find(&s[current_index..current_index + 1]) {
                Some(char_match) => sequence_start += char_match + 1,
                None => max_sequence_len = cmp::max(max_sequence_len, sequence.len() + 1),
            }
        }
        
        return max_sequence_len as i32;
    }
}